openlibrary/core/bookshelves.py

```diff
@@ -202,6 +203,101 @@ def count_total_books_logged_by_user_per_shelf(
         return {i['bookshelf_id']: i['count'] for i in result} if result else {}
 
+    # Iterates through a list of solr docs, and for all items with a 'logged edition'
+    # it will remove an item with the matching edition key from the list, and add it to
+    # doc["editions"]["docs"]
+    def link_editions_to_works(solr_docs):
+        linked_docs: list[web.storage] = []
+        editions_to_work_doc = {}
+        # adds works to linked_docs, recording their edition key and index in docs_dict if present.
+        for doc in solr_docs:
+            if doc["key"].startswith("/works"):
+                linked_docs.append(doc)
+                if doc.get("logged_edition"):
+                    editions_to_work_doc.update({doc["logged_edition"]: doc})
+        # Attaches editions to the works, in second loop-- in case of misperformed order.
+        for edition in solr_docs:
+            if edition["key"].startswith("/books/"):
+                if work_doc := editions_to_work_doc.get(edition["key"]):
+                    work_doc.editions = [edition]
+                else:
+                    # raise error no matching work found? Or figure out if the solr queries are performedorderless.
+                    logger.warning(
+                        "Error: No work found for edition %s" % edition["key"]
+                    )
+        return linked_docs
+
+    @classmethod
+    def add_storage_items_for_redirects(
+        cls, reading_log_keys, solr_docs: list[web.Storage]
+    ) -> list[web.storage]:
+        """
+        Use reading_log_keys to fill in missing redirected items in the
+        the solr_docs query results.
+
+        Solr won't return matches for work keys that have been redirected. Because
+        we use Solr to build the lists of storage items that ultimately gets passed
+        to the templates, redirected items returned from the reading log DB will
+        'disappear' when not returned by Solr. This remedies that by filling in
+        dummy works, albeit with the correct work_id.
+        """
+
+        from openlibrary.plugins.worksearch.code import run_solr_query
+        from openlibrary.plugins.worksearch.schemes.works import WorkSearchScheme
+
+        fetched_keys = {doc["key"] for doc in solr_docs}
+        missing_keys = {work for (work, _) in reading_log_keys} - fetched_keys
+
+        """
+        Provides a proper 1-to-1 connection between work keys and edition keys; needed, in order to fill in the appropriate 'logged_edition' data
+        for the correctly pulled work later on, as well as to correctly update the post-redirect version back to the pre_redirect key.
+        Without this step, processes may error due to the 'post-redirect key' not actually existing within a user's reading log.
+        """
+        work_to_edition_keys = {
+            work: edition for (work, edition) in reading_log_keys if edition
+        }
+        edition_to_work_keys = {
+            edition.split("/")[2]: work
```

Comment:
It would be nice to keep the edition/work keys in the same format throughout the code ; we generally prefer the `/works/OL123W` format; so use that throughout, and only at the very end when construction the solr query change the format ; since it's only the solr `edition_key` field that's an oddball!
```suggestion
            edition: work
```

Reply:
Change implemented.